src/bigquery.ts

```diff
@@ -2114,22 +2161,183 @@ export class BigQuery extends Service {
+        if (res.pageToken) {
+          this.trace_('[runJobsQuery] has more pages');
+          options.pageToken = res.pageToken;
+        } else {
+          this.trace_('[runJobsQuery] no more pages');
+        }
+        job!.getQueryResults(options, callback as QueryRowsCallback);
         return;
       }
-      // The Job is important for the `queryAsStream_` method, so a new query
-      // isn't created each time results are polled for.
-      options = extend({job}, queryOpts, options);
+      if (options.timeoutMs) {
+        delete options.timeoutMs;
+      }
+      this.trace_('[runJobsQuery] job not complete');
       job!.getQueryResults(options, callback as QueryRowsCallback);
     });
   }
 
+  /**
+   * Check if the given Query can run using the `jobs.query` endpoint.
+   * Returns a bigquery.IQueryRequest that can be used to call `jobs.query`.
+   * Return undefined if is not possible to convert to a bigquery.IQueryRequest.
+   *
+   * @param query string | Query
+   * @param options QueryOptions
+   * @returns bigquery.IQueryRequest | undefined
+   */
+  private buildQueryRequest_(
+    query: string | Query,
+    options: QueryOptions
+  ): bigquery.IQueryRequest | undefined {
+    if (process.env.FAST_QUERY_PATH === 'DISABLED') {
+      return undefined;
+    }
+    const queryObj: Query =
+      typeof query === 'string'
+        ? {
+            query: query,
+          }
+        : query;
+    this.trace_('[buildQueryRequest]', query, options, queryObj);
+    // This is a denylist of settings which prevent us from composing an equivalent
+    // bq.QueryRequest due to differences between configuration parameters accepted
+    // by jobs.insert vs jobs.query.
+    if (
+      !!queryObj.destination ||
+      !!queryObj.tableDefinitions ||
+      !!queryObj.createDisposition ||
+      !!queryObj.writeDisposition ||
+      (!!queryObj.priority && queryObj.priority !== 'INTERACTIVE') ||
+      queryObj.useLegacySql ||
+      !!queryObj.maximumBillingTier ||
+      !!queryObj.timePartitioning ||
+      !!queryObj.rangePartitioning ||
+      !!queryObj.clustering ||
+      !!queryObj.destinationEncryptionConfiguration ||
+      !!queryObj.schemaUpdateOptions ||
+      !!queryObj.jobTimeoutMs ||
+      // User has defined the jobID generation behavior
+      !!queryObj.jobId
+    ) {
+      return undefined;
+    }
+
+    if (queryObj.dryRun) {
+      return undefined;
+    }
+
+    if (options.job) {
+      return undefined;
+    }
+    const req: bigquery.IQueryRequest = {
+      useQueryCache: queryObj.useQueryCache,
+      labels: queryObj.labels,
+      defaultDataset: queryObj.defaultDataset,
+      createSession: queryObj.createSession,
+      maximumBytesBilled: queryObj.maximumBytesBilled,
+      timeoutMs: options.timeoutMs,
+      location: queryObj.location || options.location,
+      formatOptions: {
+        useInt64Timestamp: true,
+      },
+      maxResults: queryObj.maxResults || options.maxResults,
```

Comment:
I couldn't tell from a quick glance, do we have issues here conflating max results (e.g. more akin to a query LIMIT clause) vs the pagesize meaning used within the API?  Python has complexity around this.